### PR TITLE
Sphere surface mesh generator

### LINKED ISSUE #8414: Generated spherical surface mesh
### Description of the enhancement or error report
At the CHiMaD workshop interest was expressed to have a mesh class to generate a spherical surface mesh included in the phase field module. I implemented one in https://github.com/dschwen/CHiMaD_Hackathon

### Rationale for the enhancement or information for reproducing the error
The spherical surface is the one other domain on which phase field simulations are routinely run (vesicle surfaces)

### Identified impact
Adds a small object that makes running (and demoing) such simulations much easier.

### PR DESCRIPTION
**Pull request by dschwen**
Based on an icosahedron refinement.

sphere

Closes #8414

### CODE REVIEW

File: large_media [hunk 1]
```diff
@@ -1 +1 @@
-Subproject commit c946e65a245948f384835f05a88e6b9a13a1a090
+Subproject commit e683e3dcf9540cd17340984b6dadd5459687af16
```
[the remark this answers is not in the document]
Comment (dschwen) [reply large_media]: Yes, indeed. For the animated gif that I also added to the doc page.

**Comment by dschwen** [comment]
Search in the PR docs site seems to be broken @aeslaughter (punts you to the main site). So here's the link to the doc page:

https://mooseframework.inl.gov/docs/PRs/12991/site/source/mesh/SphereSurfaceMesh.html

**Comment by permcody** [comment]
Looks cool! Do people perform grain growth simulations on a surface sphere? Would love to do a GrainTracker run on that mesh.

**Comment by dschwen** [comment]
Oh, and that hideously large animation is actually formatted to float on the right and to be much smaller, but those style directives are currently not processed by moosedocs (seems to be a regression).

**Comment by dschwen** [comment]
> Do people perform grain growth simulations on a surface sphere

I have not seen that. But for vesicle membranes particle tracking might be needed. Nothing prevents you from just running it to test it though.

**Comment by dschwen** [comment]
I updated `large_media` master und pointed the submodule to origin/master, but it was previously pointed to an unmerged branch containing Andrew's latex test files. I have pulled those into master now. Fixing the remaining build errors as well...

**Review by friedmud — CHANGES_REQUESTED** [review]
We are no longer accepting new `Mesh` objects like this.  This should go in as a `MeshGenerator` derived object (shouldn't be much work to switch it over).

**Comment by dschwen** [comment]
@friedmud do I need to issue a `mesh->prepare_for_use(/*skip_renumber =*/false);` in the generator?

**Comment by permcody** [comment]
No

**Comment by dschwen** [comment]
Ok, I'll remove it

File: modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md [hunk 2]
```diff
@@ -0,0 +1,22 @@
...
+
+!syntax description /Mesh/SphereSurfaceMeshGenerator
+
+This object generates a 2D mesh approximating the surface of a sphere in 3D space
+using TRI3 triangle elements. Both center and radius of the sphere may be specified.
+
+!media phase_field/sphere.gif style=width:30%;margin-left:20px;float:right;
+       caption=Targeted mesh refinement prior to nucleus insertion.
+
+The mesh is constructed by iterative definement of an initial icosahedron (depth 0)
```
Comment (friedmud) [inline modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md]:
```suggestion
The mesh is constructed by iterative refinement of an initial icosahedron (depth 0)
```

**Review by friedmud — DISMISSED** [review]
I like it - one thing though: wouldn't this work with any "surface mesh"?  i.e. the same strategy should work if you start with a "cube" (not a solid cube - but a cube made out of six quad4s).

Would be cool if this object were extended to do that someday...

**Comment by dschwen** [comment]
Yeah, starting from a cube works. That was the first version I wrote. But the elements near the corners get pretty distorted and the element volumes vary write a bit.

**Comment by friedmud** [comment]
Why are we getting a large_media update here?

File: modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md (hunk 2 above)
Comment (friedmud) [reply modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md]: Wait - are you saying "definement" is actually what you want?

**Comment by dschwen** [comment]
> Why are we getting a large_media update here?

See above:

image

**Comment by permcody** [comment]
You just took a screenshot of your comment! Haha!

File: modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md (hunk 2 above)
Comment (dschwen) [reply modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md]: Uh.. no. Did I mess up accepting the suggestion? It is refinement for sure. Thanks for spotting that typo.

Comment (dschwen) [reply modules/phase_field/doc/content/source/meshgenerators/SphereSurfaceMeshGenerator.md]: Ok, "refinement" is what ended up in ac746d1. I combined the accepted suggestion commit and my fixes and tagged you as a co-author of that commit.

**Comment by dschwen** [comment]
img_20190306_095524230

Seemed like the quickest way...

File: modules/phase_field/src/meshgenerators/SphereSurfaceMeshGenerator.C [hunk 3]
```diff
@@ -0,0 +1,109 @@
...
+    _radius(getParam<Real>("radius")),
+    _center(getParam<Point>("center")),
+    _depth(getParam<unsigned int>("depth"))
+{
+}
+
+std::unique_ptr<MeshBase>
+SphereSurfaceMeshGenerator::generate()
+{
+  std::unique_ptr<ReplicatedMesh> mesh = libmesh_make_unique<ReplicatedMesh>(comm(), 2);
```
Comment (permcody) [inline modules/phase_field/src/meshgenerators/SphereSurfaceMeshGenerator.C]: This isn't the right line of code. This won't work with DistributedMesh runs or other `[Mesh]` block overrides. Do this instead:
https://github.com/idaholab/moose/blob/next/framework/src/meshgenerators/GeneratedMeshGenerator.C#L105-L106

File: modules/phase_field/test/tests/mesh/tests [hunk 4]
```diff
@@ -4,12 +4,30 @@
...
+
+  [./sphere]
+    type = 'Exodiff'
+    input = 'sphere.i'
+    cli_args = '--mesh-only'
+    exodiff = 'sphere_in.e'
+    requirement = "The mesh generator for 2d sphere surfaces in 3d space shall produce a correct mesh"
+    design = 'mesh/SphereSurfaceMesh.md'
+    issues = '#8414'
+    mesh_mode = 'REPLICATED'
```
Comment (permcody) [inline modules/phase_field/test/tests/mesh/tests]: I believe this generator will work fine with DistributedMesh when you make the other change I pointed out.

File: modules/phase_field/src/meshgenerators/SphereSurfaceMeshGenerator.C [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    {
+      Node & node = **it;
+      node = sphere.closest_point(node);
+    }
+  }
+
+  // Flatten the AMR mesh to get rid of inactive elements
+  MeshTools::Modification::flatten(*mesh);
+
+  return dynamic_pointer_cast<MeshBase>(mesh);
```
Comment (permcody) [inline modules/phase_field/src/meshgenerators/SphereSurfaceMeshGenerator.C]: I guess I didn't see this before, but we should be able to return a static_pointer_cast. MeshBase is always the base of Replicated and Distributed meshes.

**Review by permcody — CHANGES_REQUESTED** [review]

**Comment by dschwen** [comment]
Thanks for your comments @permcody . I'll make those changes. I looked at the existing code in `CartesianMeshGenerator` and `SpiralAnnularMeshGenerator` when I made my change from Mesh to MeshGenerator. I guess those will have to be updated as well.

**Comment by permcody** [comment]
Right - I haven't had a chance to go through and fix all of the existing Generators that Leora created. Many of them are still incorrect and we need to find the ones that will work with either Mesh mode, update them, fix the tests and put up PRs. If only we all had a few extra hours in the day.

**Comment by dschwen** [comment]
`static_pointer_cast` is only implemented for `shared_ptr` not `unique_ptr`...

**Comment by dschwen** [comment]
And there is more to this. With just your change I now get

```
Assertion `_mesh.is_prepared() || _mesh.is_replicated()' failed.
```

So I guess I'll have to explicitly check if the mesh is distributed and sprinkle `prepare_for_use` back in...

**Comment by dschwen** [comment]
Yeah, preparing made it not crash, but a distributed mesh generated in parallel now exodiffs

```
  FILE 1: /Users/schwd/Programs/moose/modules/phase_field/test/tests/mesh/sphere_in.e
   Title: sphere_in.e
          Dim = 3, Blocks = 1, Nodes = 162, Elements = 320, Nodesets = 0, Sidesets = 0
          Vars: Global = 0, Nodal = 0, Element = 0, Nodeset = 0, Sideset = 0, Times = 0

  FILE 2: /Users/schwd/Programs/moose/modules/phase_field/test/tests/mesh/gold/sphere_in.e
   Title: sphere_in.e
          Dim = 3, Blocks = 1, Nodes = 162, Elements = 320, Nodesets = 0, Sidesets = 0
          Vars: Global = 0, Nodal = 0, Element = 0, Nodeset = 0, Sideset = 0, Times = 0

   z coord abs diff: -4.0260323e-01 ~  6.4026032e+00 = 6.80521e+00 (node 1)
exodiff: ERROR: .. Connectivities in block id 0 are not the same.
                  First difference is node 1 of local element 1
exodiff: ERROR: .. Differences found in mesh metadata.  Aborting...
```

🙄

**Comment by permcody** [comment]
Hmm - well it's nice to have the distributed capability but if it's not going to be robust, feel free to restrict it. Not sure why it failed without an explicit prepare. That should also work just fine...

**Review by permcody — APPROVED** [review]

**Pull request merged**